### ISSUE TITLE
Duplicate the behaviour of Git merge for simple renaming / movement of files with minor editing on one branch and outright deletion on the other.

### ISSUE THREAD

**Issue by sageserpent-open** [sageserpent-open/kineticMerge#4]
This is similar to #3 , only this time the file is completed deleted on one branch - but is still renamed and / or moved on the other, possibly with some minor editing.

This is handled by ``git merge`` nicely:

1. For a conflict between a renamed / moved file *without edits* and a deletion, it writes a stage entry for 'our' version of the file only. ``git status --porcelain`` says ``AU`` for the renamed / moved file.
2. For a conflict between an *edited* and renamed / moved file and a deletion, it writes stage entries for the common base form of the file, 'our' version of the file, but leaves the conflicting deletion implied as there is neither a path nor any content for the deleted form. ``git status --porcelain`` says ``AU`` for the renamed / moved file.

Out of scope for this is what happens when a file split on one branch (possibly with edits in either part) and deleted outright on another. That is for another story.

**Comment by sageserpent-open** [comment]
Regarding:

> 1. For a conflict between a renamed / moved file _without edits_ and a deletion, it writes a stage entry for 'our' version of the file only. `git status --porcelain` says `AU` for the renamed / moved file.

A new test `com.sageserpent.kineticmerge.MainTest.outrightDeletionPropagatingThroughAFileMove` has been added in 2185d40b19f7ed00f7b020f82739da7f278e5b47.

The behaviour this test confirms **differs from what Git does**.

Kinetic Merge propagates the deletion of the file's content through the file rename / movement, resulting in the new file being empty.

Now, in the light of #33, I think that is a correct thing to do, only to be fully consistent with that ticket, it should probably report this as a `UD` or `DU` for the merged file, depending on whether the rename / movement was performed on our or their branch, respectively.

**Comment by sageserpent-open** [comment]
Regarding: 

> 2\. For a conflict between an _edited_ and renamed / moved file and a deletion, it writes stage entries for the common base form of the file, 'our' version of the file, but leaves the conflicting deletion implied as there is neither a path nor any content for the deleted form. `git status --porcelain` says `UD` for the renamed / moved file.

A new test `MainTest.outrightDeletionPropagatingThroughAnEditedFileMove` has been added in 15a451889406a5ddf680205688dd5f21a6e054a1.

Again, the  this test confirms differs from what Git does. It also isn't consistent with ticket #33, in that it should be reporting a `UD` or `DU`.

I wasn't so bothered by this for the first situation, but this one needs to be more upfront about the conflict - right now it just quietly merges just the edit changes into the renamed / moved file (the unchanged content having been deleted) and says "All's well.".

**Comment by sageserpent-open** [comment]
The desired state of affairs in a perfect world is expressed by the new tests in b880420dfeea65c944dfe18155adaf47b061e517 - these check the index.

What has been merged on to `main` is in 0de5e9781ca7fb157500290cdf2a5bdbd4951494 - this reflects the current state of affairs where in both cases the merge goes to completion without staged conflicts, and the merged renamed / moved file is either empty or is just the edited content.

The reason for climbing down on the expectations is that (unlike in #33), there is no obvious way of detecting the conflict. It is obvious from the point of view of a file whose path hasn't changed that one side has made a deletion and the other has made an edit, but if the path changes to something new, all we really know is that a file has been added.

Now Git is smart about this and does entire-file rename detection, so it can *infer* the correlation between the previous path (where the deletion is apparent) and the new path, and thus migrate the `UD` or `DU` to the new path.

Kinetic Merge on the other hand has to cope with arbitrary code motion, possibly involving splits, so it can only make that inference if the moves into the new path are **all* from the same source, namely the old path that was deleted. Actually, that can be generalised to say that as long as all the incoming moves into the new path are from paths that are deleted (and on the same side), then a deletion can be propagated as either a `UD` or `DU`.

Hmm, now I've written this, perhaps it's worth pursuing after all ... but I'm not sure that full generalisation is worth it though.

**Comment by sageserpent-open** [comment]
There is a nastier case where the new path appears on both sides as a rename / move of different paths on the base side, and both of the original paths are deleted on the opposite side. Is that a `DD`, then? Um, ah...

**Comment by sageserpent-open** [comment]
While we're on the subject of status from `git status --short`, merging a divergent renames / moves of a file from the left and right branch causes Git to report a conflict with `AU` and `UA` statuses for the two destination paths, and a `DD` for the missing original path.

I'm fairly sure that Kinetic Merge will just get on and complete the merge, reporting a divergent move of content. This seems reasonable, but is worth testing...

**Comment by sageserpent-open** [comment]
More fun and games with hypothetical situations - if a small section of a file `A.txt` is moved into a new file `B.txt` on the left, being surrounded by a massive amount of new content, and if `A.txt` is deleted outright on the right, does that really mean that `B.txt` should be regarded as `UD` in the merge?

**Comment by sageserpent-open** [comment]
Manual testing of the divergent rename use-case shows a nice outcome with Kinetic Merge - the merge goes through cleanly, with the diverging destination paths both being present. The original path is noted as being coincidentally deleted and the report shows the divergence of the content.

Tweaking this use-case by modifying the renamed file on just one branch results in just that specific path having the modification, while the other one sticks with the original content.

This is in contrast ith Git, which propagates the modification over to the other destination path and reports a conflict.

I prefer what Kinetic Merge does.

**Comment by sageserpent-open** [comment]
Come to think of it, if we're trying to duplicate what Git merge does, then the status should be `UA` or `AU` in this case - the `U` representing the *deletion* as unmerged (huh?) and the `A` representing the move / rename. Sheez.

This feels odd.

A variation is when the majority if not all of the content in the original path goes into the renamed path, but new content is put into the original path on the same side. Meanwhile the original path is deleted on the other side. Intuitively that is a rename of the original file, followed by its replacement by new content.

Git merge reports a deletion versus rename conflict with an `AU` for the destination path, but considers said path to be *unmerged* while keeping the original path.

Kinetic Merge successfully merges the destination path, propagating the deletion of content and leaves the replacement content as is at the original path.

**Comment by sageserpent-open** [comment]
Use Case 1.
--

`A.txt` is deleted entirely on the left branch.

Merge outcome for Git:
--

`A.txt` is deleted.

Merge outcome for Kinetic Merge:
--

`A.txt` is deleted.

Verdict
--

This looks good.

**Comment by sageserpent-open** [comment]
Use Case 2.
--

`A.txt` is *deleted entirely* on the left branch, and is *modified* on the right branch.

Merge outcome for Git:
--

`A.txt` remains in the worktree, marked as "DU" (deleted by us). The contents of `A.txt` are those of the right branch's version with the modification. A conflict of (modify, delete) is reported.

Merge outcome for Kinetic Merge:
--

`A.txt` remains in the worktree, marked as "DU" (deleted by us). The contents of `A.txt` are purely the modified part from the right branch, **any original content in that branch's version is deleted**.

Verdict
--

This is unsatisfactory for Kinetic Merge. I think it needs to emulate Git precisely here, because the file deletion doesn't really imply deletion of content.

**Comment by sageserpent-open** [comment]
Use Case 3.
--

`A.txt` is *deleted entirely* on the left branch, and is *renamed* to `B.txt` on the right branch.

Merge outcome for Git:
--

`A.txt` is deleted; `B.txt` remains in the worktree, marked unresolved as "UA" (added by them). The contents of `B.txt` are those of the right branch's version. A conflict of (rename, delete) is reported.

Merge outcome for Kinetic Merge:
--

`A.txt` is deleted; `B.txt` is added. The contents of `B.txt` are empty. No conflict is reported.

Verdict
--

This is **unsatisfactory** for Kinetic Merge. I think it needs to emulate Git precisely here, again because the file deletion doesn't really imply deletion of content.

**Comment by sageserpent-open** [comment]
Use Case 4.
--

`A.txt` is *deleted entirely* on the left branch, and is both *renamed* to `B.txt` and *modified* on the right branch.

Merge outcome for Git:
--

`A.txt` is deleted; `B.txt` remains in the worktree, marked unresolved as "UA" (added by them). The contents of B.txt are those of the right branch's version with the modification. A conflict of (rename, delete) is reported.

Merge outcome for Kinetic Merge:
--

`A.txt` is deleted; `B.txt` is added. The contents of `B.txt` are purely the modified part from the right branch, any original content in that branch's version is deleted. A conflict is reported.


Verdict
--

This is **unsatisfactory** for Kinetic Merge. I think it needs to emulate Git precisely here, because the file deletion doesn't really imply deletion of content.

**Comment by sageserpent-open** [comment]
Use Case 5.
--

`A.txt` is *replaced with new content* on the left branch, and is *renamed* to `B.txt` on the right branch.

Merge outcome for Git:
--

`A.txt` is deleted; `B.txt` is added. The contents of `B.txt` are those of the left branch's version with replacement content. No conflict is reported.

Merge outcome for Kinetic Merge:
--

`A.txt` is deleted; `B.txt` is added. The contents of `B.txt` are those of the left branch's version with replacement content. No conflict is reported.orted.


Verdict
--

This looks **good**.

**Comment by sageserpent-open** [comment]
Use Case 6.
--

`A.txt` is *replaced with new content* on the left branch, and is both *renamed* to `B.txt` and *modified* on the right branch.

Merge outcome for Git:
--

`A.txt` is deleted; `B.txt` remains in the worktree, marked unresolved as "UU" (both modified). The contents of `B.txt` are a conflict between the left and right branch versions. A conflict of (content) is reported.

Merge outcome for Kinetic Merge:
--

`A.txt` is deleted; `B.txt` is added. The contents of `B.txt` are the isolated modification of `B.txt` made on the right with a replacement of the original content in `A.txt` with that of `A.txt` on the left. No conflict is reported.


Verdict
--

This is **good** for Kinetic Merge. I think it does a better job than Git here, because it applies the replacement in the right context in `B.txt`, dovetailing it with the modification on the other side.

**Comment by sageserpent-open** [comment]
Use Case 7.
--

`A.txt` is *deleted entirely* on the left branch, and is *renamed* to `B.txt` on the right branch together with a replacement version of `A.txt`.

Merge outcome for Git:
--

`A.txt` remains in the worktree, marked unresolved as **"DU"** (deleted by us); `B.txt` is added. The contents of `A.txt` are those of the right branch's replacement version. The contents of B.txt are those of the right branch's version with the modification. A conflict of (modify, delete) is reported.

Merge outcome for Kinetic Merge:
--

`A.txt` remains in the worktree, marked unresolved as **"DU"** (deleted by us); `B.txt` is added. The contents of `A.txt` are those of the right branch's replacement version. The contents of `B.txt` are empty. A conflict is reported.

Verdict
--

This is **unsatisfactory** for Kinetic Merge. I think it needs to emulate Git precisely here, again because the file deletion doesn't really imply deletion of content.

**Comment by sageserpent-open** [comment]
Use Case 8.
--

`A.txt` is *deleted entirely* on the left branch, and is both *renamed* to `B.txt` and *modified* on the right branch together with a *replacement version* of `A.txt`.

Merge outcome for Git:
--

`A.txt` remains in the worktree, marked unresolved as **"DU"** (deleted by us); `B.txt` is added. The contents of `A.txt` are those of the right branch's replacement version. The contents of `B.txt` are those of the right branch's version with the modification. A conflict of (modify, delete) is reported.

Merge outcome for Kinetic Merge:
--

`A.txt` remains in the worktree, marked unresolved as **"DU"** (deleted by us); `B.txt` is added. The contents of `A.txt` are those of the right branch's replacement version. The contents of `B.txt` are purely the modified part from the right branch, any original content in that branch's version is deleted. A conflict is reported.

Verdict
--

This is **unsatisfactory** for Kinetic Merge. I think it needs to emulate Git precisely here, again because the file deletion doesn't really imply deletion of content.

**Comment by sageserpent-open** [comment]
Now, if instead of *deleting a file entirely*, we had simply made its content *empty*, then I'd expect Kinetic Merge to cleanly merge in all cases, and yes, that emptying of content should be migrated as a deletion.

**Comment by sageserpent-open** [comment]
Over and above playing catch-up with Git merge for entire-file deletions, Kinetic Merge also has to worry about split file renames, where the content is divided up into multiple new files (or put into existing files).

If there are multiple new files, then a straightforward generalisation would be to mark them all as **"UA"**.

If the destinations are within existing files, then again, they should be marked as **"UA"**.

However, given that we may just have the odd small section or two migrating from the outgoing file to either new or existing files, it seems reasonable to only go for an unresolved outcome if the majority of the merged destination file is composed of sections coming from a deleted file or files.

(Yes, Kinetic Merge also has to worry about multiple sections from *different* files being deleted arriving in the same destination file.)

**Comment by sageserpent-open** [comment]
For a final wrinkle, it is possible to have our`A.txt` be split and divergently renamed across the left and right branches.

Here, the one half of `A.txt` goes to `B.txt` on the left and the other half goes to `C.txt` on the right.

The desirable outcome would be for both renames to go into the merge as is, as they involve disjoint sections.

**Comment by sageserpent-open** [comment]
PLAN
--
1. Work on deletion versus rename conflicts first - that is what this ticket is about, after all. For the side that deletes a file entirely, simply omit its path when building the sources for that side. It has gone. This has the effect of turning a straight rename into a pure move of just one section (bar incidental smaller ambiguous matches elsewhere) and of turning rename plus edits into smaller pure moves with the original content being edited simply being dropped. Once this is done (in `CodeMotionAnalysisExtension.merge`), the caller can check the original path, looking for sections in the base `Sources` instance that have migrated to another or several other paths. If more than half of the content (counting in tokens) at the original path *and* destination path has migrated in this way, Kinetic Merge declares a (rename, delete) conflict.

**Comment by sageserpent-open** [comment]
> 1. There are some weasel words here - we may have multiple moves to _different_ paths; as long as they all sum up to more than half of the original path's content, this is considered to be grounds for a conflict on each destination path. More on that later...

Consider the use case where a file `A.txt` is split into two halves and they are moved to different paths, but on either side of the merge. Thus `FirstHalf.txt` on the left and `SecondHalf.txt` on the right. `A.txt` is deleted entirely on both the left and right.

As mentioned above, the desirable outcome is to have both `FirstHalf.txt` and `SecondHalf.txt` appear in the merge, with no mention of conflicts as all of `A.txt` has been spoken for by the combination of the moves.

So this has to be dealt with as a special case.

This leads to a fold within the wrinkle ... what happens if `A.txt` is split into thirds and only two thirds move to distinct paths; the remaining third being dropped by virtue of `A.txt` being deleted entirely on both the left and right. Does that mean that we're back to declaring conflicts again?

**Comment by sageserpent-open** [comment]
Couple of things to spell out:

1. Whether or not there is a rename versus deletion conflict is associated with the *original* path, although it is reported on the destination paths. So if there are several original paths whose content feeds into some shared destination path and only one of the them is considered conflicted, that will mark the destination path as conflicted (albeit the report should reference the original path in question).
2. If a base section from the original path is moved on **both the left and right** side, then regardless of whether it moves divergently or coincidentally or with ambiguity, we have accounted for the base section and thus it does not cause a conflict wrt the original path. Obviously we may have multiple base sections moving to distinct or shared destinations, so we have to consider all destination paths and coalesce the outcomes for each path.

**Comment by sageserpent-open** [comment]
Got something going for the third and fourth use cases above, the tests pass.

There is a known deficiency in that a deletion versus renaming conflict can be declared when content from a coincidentally deleted file moves on one side to an *existing* file, need to sort that out.

Also need more tests for the other use cases.

**Comment by sageserpent-open** [comment]
Tested for and fixed the deficiency in 30097efa316e7e37dde438fff0d8eae880a726f6.

Now for the other use cases...

**Comment by sageserpent-open** [comment]
Added a test for the seventh use case in 8afe1b58b2d62e3ca2d34f58983f2284872e054c, and this fails. It does detect the desired reporting of a conflict in the file playing the role of `A.txt`, but `B.txt` is empty.

In other words, the deletion of `A.txt` is still being treated as a content deletion to be migrated to `B.txt`. This is not surprising, as the work doen so far has focussed on coincident deletions of a path...

**Comment by sageserpent-open** [comment]
Fixed the outstanding test failure for the seventh use case in d89b32760dcb41a2a5cd3ca40198bc0bcf094186. The implementation is a bit of a cryptic mess, though.

**Comment by sageserpent-open** [comment]
Added a test for the eighth use case in db29f5e91c3787c6ec70203a2f463b77548c8b66, naturally it fails.

What is unexpected is that the new, cryptic and messy implementation also fails this test in the merge 85aeb19de78d6fdcc947f5c80954bbffdf1d31f1.

Maybe I shouldn't be so surprised...

**Comment by sageserpent-open** [comment]
This last failure is down to the further editing of the renamed file causing matches to become smaller, as there is an alien fragment introduced by the edit. That in turn causes only the first match to be interpreted as a deletion versus edit conflict, which is treated correctly. The rest of the matches results in coincident deletions, and these are treated as migrated deletions.

Need to take a step back and think about whether this approach is the right one...

**Comment by sageserpent-open** [comment]
Tried an alternative approach, culminating in 467f79f2ce57473ddb05e2e286bfe50bdf634338. This also fails the test for the eight use case, and in exactly the same way.

At least the implementation is less cryptic.

**Comment by sageserpent-open** [comment]
All tests are passing in `MainTest` as of 3f6a6cb48f98a5410e70759302dcbd2e99183a41.

The example merge from #35 is looking as good as ever.

**Comment by sageserpent-open** [comment]
One last pass over this for additional test to cover any straggler scenarios discussed above, then hopefully it's on to #5 and #137...

**Comment by sageserpent-open** [comment]
> There is a nastier case where the new path appears on both sides as a rename / move of different paths on the base side, and both of the original paths are deleted on the opposite side. Is that a `DD`, then? Um, ah...
> 
> ... Tried it with Git, it thinks it's an `AA` and preserves the content of both of the original paths as a big conflict, refusing to acknowledge the two deletions.

Tried it with Kinetic Merge as of 3f6a6cb, got the same outcome with the same *manual* test.

There is however no mention of a (rename/rename) conflict, which Git does inform the user about.

So let's add an automated test to document this behaviour and add in some diagnostic...

**Comment by sageserpent-open** [comment]
Added the new test in bce015be129ab84859ce6a2407660820e77b2cd1; it passes, duplicating the behaviour of Git.

I can't be bothered to replicate the Git (rename/rename) conflict message for now.

**Comment by sageserpent-open** [comment]
Branch `main` fast-forwarded to bce015be129ab84859ce6a2407660820e77b2cd1. Ship it!

**Comment by sageserpent-open** [comment]
> I can't be bothered to replicate the Git (rename/rename) conflict message for now.

Guilt prevailed, thus there are additional diagnostics for detecting renames in the converging renames use-case, as well as for straight renames, splits and migration of edits and / or deletions through renames.

I still haven't mirrored the Git (rename/rename) conflict message, and I can live with that shame for now.
